### ISSUE TITLE
"Clipping" predictions

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/multidms#13]
While clipping the functional score targets seems to be quite beneficial for the model in a number of ways -- @WSDeWitt raised a very good point -- that the g(z) sigmoidal function is limiting us by learning the lower bound of clipped data. 

![image](https://user-images.githubusercontent.com/23087427/194605464-ee805fa1-f7a4-40fa-b120-3da8231e0520.png)

The reason this is so limiting in our current setup is because [our sigmoid](https://github.com/matsengrp/multidms/blob/b06689275e6bfdf410f8824a361d180b1576b8d1/multidms/model.py#L100) is only flexible by two parameters stretch, and shift (both in the Y direction.) $g_{\alpha_{stretch, shift}}(z) = (\alpha_{stretch} / (1 + s^{-x})) + \alpha_{shift}$

While this a simple and effective solution, the symmetric nature of the sigmoid limits us. This is because a large portion of the data we train on is bias towards more deleterious (negative functional score) examples - and thus the model tries hard to fit the sigmois when $z <= 0$, and the other half of the sigmoid is constrained to mirror that shape on the other side. How might we allow for our model to more effectively fit interesting patterns on the upper (higher functional score variants) end of the $g(z)$ function?  

There are a few solutions to handling this that could be used alone or in conjunction:

1. We could have a more complex (more parameters) $g(z)$, but this adds potentially unnecessary complexity and may introduce over-fitting as well as break some of the nice properties of such a simple sigmoid. This may be another optional model, but not ideal for our purposes.
2. Tossing data below the clipping threshold could be beneficial, but also makes us loose data informing the single mutational effects in the data (Beta's) - and it wouldn't solve the problem of over-fitting to a still large majority of deleterious mutations.
3. We clip the prediction of the model. Given that we already know the bounds of prediction when clipping, we don't need the model to learn this. In essence, the model may want to stretch of shift but is limited because it encounters higher loss when the desired shpe does not conform to our artificial threshold. 

Per discussion, it seems option #3 is the way to go. By adding another activation to the output of out prediction function, we essentially give the model a perfect fit on variants below the threshold for free, and let it fit it's shape to the rest of the data. With a simple $g_{\alpha}(\phi(x))$ model, this would essentially boil down to wrapping the function with another function, $t()$ that applies a smooth relu activation to the output of the model where all predictions are floored at the lower threshold of the clip. Then our model would look like $f(x) = t(g_{\alpha}(\phi(x)))$. 

In the objective function, we've added a [softmax](https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.softplus.html), where the "smooth hinge" sits at (-lower bound, -lower bound) and thus the model incurs 0 loss when predicting below the clipped threshold. Concretely, we applied 

$$
t(p, l) = log(1+e^{p-l}) + l
$$

where $p$ is the output of $f(v, h)$, and $l$ is a constant value to change the position of the softmax "hinge". 

After [implementing this](https://github.com/matsengrp/multidms/blob/1a555c098dbf39437f5dd585173ffa6846671c39/multidms/model.py#L159), the respective change to the fit above is:
 
![image](https://user-images.githubusercontent.com/23087427/194608567-a6583dca-1da0-4931-a467-6f679cf3a2fe.png)

It seems to have been helpful. We can see the model no longer was constrained top fitting the clipping threshold, and may have even fit more fairly to the Omicron data. @WSDeWitt Any thoughts? Is this what you had in mind?
